File: windows_agent/browsers.py
"""Chrome and Edge browsing history extraction.

Reads the SQLite History database from each browser's user profile.
The DB is copied to a temp file first since browsers hold a lock on it.
"""

import logging
import os
import shutil
import sqlite3
import tempfile
from datetime import datetime, timedelta, timezone
from typing import List, NamedTuple

logger = logging.getLogger("browser_reporter_agent")

# Chrome/Edge use WebKit timestamps: microseconds since 1601-01-01 00:00:00 UTC
_WEBKIT_EPOCH_DIFF = 11644473600  # seconds between 1601-01-01 and 1970-01-01


class Visit(NamedTuple):
    url: str
    title: str
    visit_time_webkit: int  # raw WebKit timestamp for state tracking
    visit_time_epoch_ms: int  # epoch milliseconds for API


def webkit_to_epoch_ms(webkit_ts: int) -> int:
    """Convert Chrome WebKit timestamp to Unix epoch milliseconds."""
    return int((webkit_ts / 1_000_000 - _WEBKIT_EPOCH_DIFF) * 1000)


def epoch_ms_to_webkit(epoch_ms: int) -> int:
    """Convert Unix epoch milliseconds to Chrome WebKit timestamp."""
    return int((epoch_ms / 1000 + _WEBKIT_EPOCH_DIFF) * 1_000_000)


def _max_history_webkit(max_days: int) -> int:
    """Return the WebKit timestamp for max_days ago."""
    cutoff = datetime.now(timezone.utc) - timedelta(days=max_days)
    epoch_s = cutoff.timestamp()
    return int((epoch_s + _WEBKIT_EPOCH_DIFF) * 1_000_000)


def _find_history_paths(browser_data_dir: str) -> List[str]:
    """Find all History database files across browser profiles."""
    paths = []
    if not os.path.isdir(browser_data_dir):
        return paths

    # Default profile
    default = os.path.join(browser_data_dir, "Default", "History")
    if os.path.isfile(default):
        paths.append(default)

    # Additional profiles (Profile 1, Profile 2, etc.)
    for entry in os.listdir(browser_data_dir):
        if entry.startswith("Profile "):
            p = os.path.join(browser_data_dir, entry, "History")
            if os.path.isfile(p):
                paths.append(p)

    return paths


def _query_history(db_path: str, since_webkit: int) -> List[Visit]:
    """Copy the History DB to temp and query visits since the given timestamp."""
    visits: List[Visit] = []
    tmp_path = None
    try:
        # Copy to temp — browser locks the original
        fd, tmp_path = tempfile.mkstemp(suffix=".sqlite")
        os.close(fd)
        shutil.copy2(db_path, tmp_path)

        conn = sqlite3.connect(f"file:{tmp_path}?mode=ro", uri=True)
        try:
            cursor = conn.execute(
                """
                SELECT urls.url, urls.title, visits.visit_time
                FROM visits
                INNER JOIN urls ON visits.url = urls.id
                WHERE visits.visit_time > ?
                ORDER BY visits.visit_time ASC
                """,
                (since_webkit,),
            )
            for url, title, vt in cursor.fetchall():
                visits.append(Visit(
                    url=url or "",
                    title=title or "",
                    visit_time_webkit=vt,
                    visit_time_epoch_ms=webkit_to_epoch_ms(vt),
                ))
        finally:
            conn.close()
    except Exception as e:
        logger.error("Failed to read history from %s: %s", db_path, e)
    finally:
        if tmp_path and os.path.isfile(tmp_path):
            try:
                os.remove(tmp_path)
            except OSError:
                pass

    return visits


def collect_visits(
    browser: str,
    since_webkit: int,
    max_history_days: int,
) -> List[Visit]:
    """Collect visits from a browser since the given WebKit timestamp.

    Args:
        browser: "chrome" or "edge"
        since_webkit: WebKit timestamp of last sent visit (0 for all)
        max_history_days: Maximum age of visits to collect
    """
    local_app_data = os.environ.get("LOCALAPPDATA", "")
    if not local_app_data:
        logger.error("LOCALAPPDATA not set")
        return []

    if browser == "chrome":
        data_dir = os.path.join(local_app_data, "Google", "Chrome", "User Data")
    elif browser == "edge":
        data_dir = os.path.join(local_app_data, "Microsoft", "Edge", "User Data")
    else:
        logger.error("Unknown browser: %s", browser)
        return []

    # Use the more recent of since_webkit and max_history_days cutoff
    cutoff = _max_history_webkit(max_history_days)
    effective_since = max(since_webkit, cutoff)

    history_paths = _find_history_paths(data_dir)
    if not history_paths:
        logger.info("No %s history databases found", browser)
        return []

    all_visits: List[Visit] = []
    for path in history_paths:
        logger.info("Reading %s history: %s", browser, path)
        all_visits.extend(_query_history(path, effective_since))

    logger.info("Collected %d visits from %s", len(all_visits), browser)
    return all_visits

File: windows_agent_go/browsers.go
package main

import (
	"database/sql"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"

	_ "modernc.org/sqlite"
)

// WebKit timestamps: microseconds since 1601-01-01 00:00:00 UTC
const webkitEpochDiff = 11644473600 // seconds between 1601-01-01 and 1970-01-01

// Visit represents a single browser history entry.
type Visit struct {
	URL             string
	Title           string
	WebkitTimestamp int64  // raw WebKit timestamp for state tracking
	EpochMS         int64  // Unix epoch milliseconds for the API
	Browser         string // "chrome" or "edge"
	Profile         string // "Default", "Profile 1", etc.
}

func webkitToEpochMS(webkitTS int64) int64 {
	return (webkitTS/1_000_000 - webkitEpochDiff) * 1000
}

// maxHistoryWebkit returns the WebKit timestamp for maxDays ago.
func maxHistoryWebkit(maxDays int) int64 {
	cutoff := time.Now().UTC().Add(-time.Duration(maxDays) * 24 * time.Hour)
	epochSec := cutoff.Unix()
	return (epochSec + webkitEpochDiff) * 1_000_000
}

// browserDataDir returns the User Data directory for a browser.
func browserDataDir(browser string) string {
	localAppData := os.Getenv("LOCALAPPDATA")
	if localAppData == "" {
		return ""
	}
	switch browser {
	case "chrome":
		return filepath.Join(localAppData, "Google", "Chrome", "User Data")
	case "edge":
		return filepath.Join(localAppData, "Microsoft", "Edge", "User Data")
	default:
		return ""
	}
}

// profileEntry holds the profile directory name and History path.
type profileEntry struct {
	name    string // "Default", "Profile 1", etc.
	dbPath  string
}

// findHistoryPaths finds all History database files across browser profiles.
func findHistoryPaths(dataDir string) []profileEntry {
	var entries []profileEntry
	if dataDir == "" {
		return entries
	}

	// Default profile
	defaultPath := filepath.Join(dataDir, "Default", "History")
	if info, err := os.Stat(defaultPath); err == nil && !info.IsDir() {
		entries = append(entries, profileEntry{name: "Default", dbPath: defaultPath})
	}

	// Additional profiles (Profile 1, Profile 2, etc.)
	dirEntries, err := os.ReadDir(dataDir)
	if err != nil {
		return entries
	}
	for _, e := range dirEntries {
		if e.IsDir() && strings.HasPrefix(e.Name(), "Profile ") {
			p := filepath.Join(dataDir, e.Name(), "History")
			if info, err := os.Stat(p); err == nil && !info.IsDir() {
				entries = append(entries, profileEntry{name: e.Name(), dbPath: p})
			}
		}
	}
	return entries
}

// copyFile copies src to dst.
func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()
	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()
	_, err = io.Copy(out, in)
	return err
}

// queryHistory copies the History DB to a temp file and queries visits.
func queryHistory(dbPath string, sinceWebkit int64, browser, profile string) []Visit {
	tmpFile, err := os.CreateTemp("", "br_history_*.sqlite")
	if err != nil {
		logger.Printf("ERROR creating temp file: %v", err)
		return nil
	}
	tmpPath := tmpFile.Name()
	tmpFile.Close()
	defer os.Remove(tmpPath)

	if err := copyFile(dbPath, tmpPath); err != nil {
		logger.Printf("ERROR copying history db %s: %v", dbPath, err)
		return nil
	}

	db, err := sql.Open("sqlite", fmt.Sprintf("file:%s?mode=ro", tmpPath))
	if err != nil {
		logger.Printf("ERROR opening sqlite %s: %v", tmpPath, err)
		return nil
	}
	defer db.Close()

	rows, err := db.Query(`
		SELECT urls.url, urls.title, visits.visit_time
		FROM visits
		INNER JOIN urls ON visits.url = urls.id
		WHERE visits.visit_time > ?
		ORDER BY visits.visit_time ASC
	`, sinceWebkit)
	if err != nil {
		logger.Printf("ERROR querying history: %v", err)
		return nil
	}
	defer rows.Close()

	var visits []Visit
	for rows.Next() {
		var url, title sql.NullString
		var vt int64
		if err := rows.Scan(&url, &title, &vt); err != nil {
			logger.Printf("ERROR scanning row: %v", err)
			continue
		}
		visits = append(visits, Visit{
			URL:             url.String,
			Title:           title.String,
			WebkitTimestamp: vt,
			EpochMS:         webkitToEpochMS(vt),
			Browser:         browser,
			Profile:         profile,
		})
	}
	return visits
}

// CollectVisits collects all visits from a browser since the given WebKit timestamp.
func CollectVisits(browser string, sinceWebkit int64, maxDays int) []Visit {
	dataDir := browserDataDir(browser)
	if dataDir == "" {
		logger.Printf("WARN LOCALAPPDATA not set, skipping %s", browser)
		return nil
	}

	cutoff := maxHistoryWebkit(maxDays)
	if sinceWebkit > cutoff {
		cutoff = sinceWebkit
	}

	profiles := findHistoryPaths(dataDir)
	if len(profiles) == 0 {
		logger.Printf("INFO no %s history databases found", browser)
		return nil
	}

	var all []Visit
	for _, p := range profiles {
		logger.Printf("INFO reading %s history: %s (profile: %s)", browser, p.dbPath, p.name)
		visits := queryHistory(p.dbPath, cutoff, browser, p.name)
		all = append(all, visits...)
	}
	logger.Printf("INFO collected %d visits from %s", len(all), browser)
	return all
}
